Refreshing dashboard crashes the app

- SQL Operations Studio Version: 0.28.1

Steps to Reproduce: 

1. Open any dashboard. On the right most side, click on the down (or up) arrow button to hide/reveal the dashboard details. 
2. Click on the `...` to its right and click refresh.

![capture1111](https://user-images.githubusercontent.com/6411451/38385143-f9215188-38c5-11e8-99c8-d6a3eb65aa13.JPG)


Expected - Dashboard refreshes
Actual - The app crashes with the following error - 

```
[uncaught exception]: Error: ViewDestroyedError: Attempt to use a destroyed view: detectChanges

C:\Users\adbist\Desktop\SQL_Tools\sqlopsstudio\node_modules\zone.js\dist\zone-node.js:2251 Error: ViewDestroyedError: Attempt to use a destroyed view: detectChanges
    at viewDestroyedError (C:\Users\adbist\Desktop\SQL_Tools\sqlopsstudio\node_modules\@angular\core\bundles\core.umd.js:8665:12)
    at Object.debugUpdateDirectives [as updateDirectives] (C:\Users\adbist\Desktop\SQL_Tools\sqlopsstudio\node_modules\@angular\core\bundles\core.umd.js:12810:15)
    at checkAndUpdateView (C:\Users\adbist\Desktop\SQL_Tools\sqlopsstudio\node_modules\@angular\core\bundles\core.umd.js:12151:14)
    at callWithDebugContext (C:\Users\adbist\Desktop\SQL_Tools\sqlopsstudio\node_modules\@angular\core\bundles\core.umd.js:13213:42)
    at Object.debugCheckAndUpdateView [as checkAndUpdateView] (C:\Users\adbist\Desktop\SQL_Tools\sqlopsstudio\node_modules\@angular\core\bundles\core.umd.js:12753:12)
    at ViewRef_.detectChanges (C:\Users\adbist\Desktop\SQL_Tools\sqlopsstudio\node_modules\@angular\core\bundles\core.umd.js:10225:63)
    at SafeSubscriber._next (file:///C:/Users/adbist/Desktop/SQL_Tools/sqlopsstudio/out/sql/parts/dashboard/widgets/properties/propertiesWidget.component.js:57:34)
    at SafeSubscriber.__tryOrUnsub (C:\Users\adbist\Desktop\SQL_Tools\sqlopsstudio\node_modules\rxjs\Subscriber.js:238:16)
    at SafeSubscriber.next (C:\Users\adbist\Desktop\SQL_Tools\sqlopsstudio\node_modules\rxjs\Subscriber.js:185:22)
    at Subscriber._next (C:\Users\adbist\Desktop\SQL_Tools\sqlopsstudio\node_modules\rxjs\Subscriber.js:125:26)
```

--- a/src/sql/parts/dashboard/common/actions.ts
+++ b/src/sql/parts/dashboard/common/actions.ts
@@ -210,7 +210,6 @@ export class CollapseWidgetAction extends Action {
 		private _uri: string,
 		private _widgetUuid: string,
 		private collpasedState: boolean,
-		private collapsedStateChangedEvent: Event<boolean>,
 		@IAngularEventingService private _angularEventService: IAngularEventingService
 	) {
 		super(
@@ -218,7 +217,6 @@ export class CollapseWidgetAction extends Action {
 			collpasedState ? CollapseWidgetAction.EXPAND_LABEL : CollapseWidgetAction.COLLPASE_LABEL,
 			collpasedState ? CollapseWidgetAction.EXPAND_ICON : CollapseWidgetAction.COLLAPSE_ICON
 		);
-		this.collapsedStateChangedEvent(collapsed => this._updateState(collapsed));
 	}
 
 	run(): TPromise<boolean> {
@@ -232,8 +230,15 @@ export class CollapseWidgetAction extends Action {
 	}
 
 	private _updateState(collapsed: boolean): void {
+		if (collapsed === this.collpasedState) {
+			return;
+		}
 		this.collpasedState = collapsed;
 		this._setClass(this.collpasedState ? CollapseWidgetAction.EXPAND_ICON : CollapseWidgetAction.COLLAPSE_ICON);
 		this._setLabel(this.collpasedState ? CollapseWidgetAction.EXPAND_LABEL : CollapseWidgetAction.COLLPASE_LABEL);
+	}
+
+	public set state(collapsed: boolean) {
+		this._updateState(collapsed);
 	}
 }

--- a/src/sql/parts/dashboard/contents/dashboardWidgetWrapper.component.ts
+++ b/src/sql/parts/dashboard/contents/dashboardWidgetWrapper.component.ts
@@ -53,6 +53,7 @@ export class DashboardWidgetWrapper extends AngularDisposable implements OnInit 
 	@Input() private _config: WidgetConfig;
 	@Input() private collapsable = false;
 
+	private _collapseAction: CollapseWidgetAction;
 	private _collapsed = false;
 
 	public get collapsed(): boolean {
@@ -64,16 +65,12 @@ export class DashboardWidgetWrapper extends AngularDisposable implements OnInit 
 			return;
 		}
 		this._collapsed = val;
-		if (this.collapsedStateChangedEmitter) {
-			this.collapsedStateChangedEmitter.fire(this._collapsed);
-		}
+		this._collapseAction.state = val;
 		this._changeref.detectChanges();
 		if (!val) {
 			this.loadWidget();
 		}
 	}
-
-	private collapsedStateChangedEmitter: Emitter<boolean>;
 
 	@memoize
 	public get guid(): string {
@@ -114,8 +111,8 @@ export class DashboardWidgetWrapper extends AngularDisposable implements OnInit 
 		this._actionbar = new ActionBar(this._actionbarRef.nativeElement);
 		if (this._actions) {
 			if (this.collapsable) {
-				this.collapsedStateChangedEmitter = new Emitter<boolean>();
-				this._actionbar.push(this._bootstrap.instantiationService.createInstance(CollapseWidgetAction, this._bootstrap.getUnderlyingUri(), this.guid, this.collapsed, this.collapsedStateChangedEmitter.event), { icon: true, label: false });
+				this._collapseAction = this._bootstrap.instantiationService.createInstance(CollapseWidgetAction, this._bootstrap.getUnderlyingUri(), this.guid, this.collapsed);
+				this._actionbar.push(this._collapseAction, { icon: true, label: false });
 			}
 			this._actionbar.push(this._bootstrap.instantiationService.createInstance(ToggleMoreWidgetAction, this._actions, this._component.actionsContext), { icon: true, label: false });
 		}
@@ -123,13 +120,13 @@ export class DashboardWidgetWrapper extends AngularDisposable implements OnInit 
 	}
 
 	public refresh(): void {
-		if (this._component && this._component.refresh) {
+		if (!this.collapsed && this._component && this._component.refresh) {
 			this._component.refresh();
 		}
 	}
 
 	public layout(): void {
-		if (this._component && this._component.layout) {
+		if (!this.collapsed && this._component && this._component.layout) {
 			this._component.layout();
 		}
 	}
@@ -170,7 +167,7 @@ export class DashboardWidgetWrapper extends AngularDisposable implements OnInit 
 			this._component = componentRef.instance;
 			let actions = componentRef.instance.actions;
 			if (componentRef.instance.refresh) {
-				actions.push(new RefreshWidgetAction(componentRef.instance.refresh, componentRef.instance));
+				actions.push(new RefreshWidgetAction(this.refresh, this));
 			}
 			if (actions !== undefined && actions.length > 0) {
 				this._actions = actions;

